### PR TITLE
json: Improve tolerance to uncovered C members during JSON encoding

### LINKED ISSUE #50976: JSON array encoding fails on array of objects
Hi,

When I try to encode an array of objects. If the descriptors of the inner object doesn't cover all the fields of the object structure, then the array is wrongly encoded.

**Normal case:**
```c
#define MY_ARRAY_SIZE 4u

struct mystruct_obj
{
	uint32_t a;
	/* Uncomment following line to get an incorrect result */
	// uint32_t b;
	uint32_t n;
};

static const struct json_obj_descr descr_mystruct_obj[] = {
	JSON_OBJ_DESCR_PRIM(struct mystruct_obj, a, JSON_TOK_NUMBER),
	JSON_OBJ_DESCR_PRIM(struct mystruct_obj, n, JSON_TOK_NUMBER),
};

struct mystruct_arr
{
	struct mystruct_obj items[MY_ARRAY_SIZE];
	size_t count;
};

static const struct json_obj_descr descr_mystruct_arr[] = {
	JSON_OBJ_DESCR_OBJ_ARRAY(
		struct mystruct_arr,
		items,
		MY_ARRAY_SIZE,
		count,
		descr_mystruct_obj,
		ARRAY_SIZE(descr_mystruct_obj)
	)
};

int encode_to(char *buf, size_t len)
{
	struct mystruct_arr arr = { .count = 0u };

	for (uint8_t i = 0u; i < MY_ARRAY_SIZE; i++) {
		arr.items[i].a = 3;
                // Uncomment following line to highlight that the encoder reads at incorrect offsets.
		// arr.items[i].b = 0xFFFFFFFF;
		arr.items[i].n = i;
		arr.count++;
	}

	return json_arr_encode_buf(descr_mystruct_arr, &arr, buf, len);
}
```

With the previous code, the json is properly encoded :
```json
[
	{
		"a": 3,
		"n": 0
	},
	{
		"a": 3,
		"n": 1
	},
	{
		"a": 3,
		"n": 2
	},
	{
		"a": 3,
		"n": 3
	}
]
```

But if I uncomment the line `uint32_t b;`.
Here, "b" member is still not encoded in the JSON, but I would have expect the result to be the same, however I get the following result. Array is incorrectly encoded.

```json
[
	{
		"a": 3,
		"n": 0
	},
	{
		"a": 0,
		"n": 0
	},
	{
		"a": 0,
		"n": 3
	},
	{
		"a": 3,
		"n": 2
	}
]
```

If I set "b" to "0xFFFFFFFF" in initialization loop, it seems that the value of "b" is used sporadically. This highlights that the encoder reads the values at incorrect offsets while encoding the array.
```json
[
	{
		"a": 3,
		"n": 0
	},
	{
		"a": 0,
		"n": -1
	},
	{
		"a": -1,
		"n": 3
	},
	{
		"a": 3,
		"n": 2
	}
]
```

Is this just an undocummented unhandled case ? I do understand this case has not been foreseen. 

**Impact**
Impact is that we cannot use C structure (for JSON encoding) which are not exhaustively described using descriptors. In my case, I will not be able to use existing structure if there are fields I don't want to encode, I would need to create structures dedicated to JSON encoding.

I haven't done much debugging yet, maybe I will dig into this when I have time.

**Environment:**
 - Linux
 - Zephyr v3.1.0
 - QEMU x86

### PR DESCRIPTION
**Pull request by lucasdietrich**
This enhancement fixes encoding of JSON objects for which descriptors do not covers all members of the C structure represented.

Fix #50976

Signed-off-by: Lucas Dietrich <ld.adecy@gmail.com>

### CODE REVIEW

**Comment by carlescufi** [comment]
@mrfuchs could you please take a look? Thanks!

**Comment by MaureenHelm** [comment]
@M1cha can you take a look?

File: include/zephyr/data/json.h [hunk 1]
```diff
@@ -137,6 +138,9 @@ typedef int (*json_append_bytes_t)(const char *bytes, size_t len,
 				 __alignof__(type) == 2 ? 1 : \
 				 __alignof__(type) == 4 ? 2 : 3)
 
+/* Helper to get the size of a member within a struct. */
+#define Z_MEMBER_SIZE(type, member) sizeof(((type *)0)->member)
```
Comment (M1cha) [inline include/zephyr/data/json.h:142]: Given how often I needed this in other zephyr projects I'd say it's finally time to add this to `include/zephyr/sys/util.h` with the name `SIZEOF_FIELD` (that's what the linux kernel calls this, just lowercase).

File: include/zephyr/data/json.h [hunk 2]
```diff
@@ -109,7 +109,8 @@ struct json_obj_descr {
 	union {
 		struct {
 			const struct json_obj_descr *sub_descr;
-			size_t sub_descr_len;
+			uint16_t sub_descr_len;
+			uint16_t sub_struct_size;
```
Comment (M1cha) [inline include/zephyr/data/json.h:113]: what's the reason for the type change and why did you choose `uint16_t`?

File: tests/lib/json/src/main.c [hunk 3]
```diff
@@ -255,10 +264,11 @@ ZTEST(lib_json_test, test_json_limits)
 		     "Integer limits not decoded correctly");
 }
 
+struct obj_array_array obj_array_array_ts;
+
 ZTEST(lib_json_test, test_json_decoding_array_array)
 {
 	int ret;
-	struct obj_array_array obj_array_array_ts;
```
[the remark this answers is not in the document]
Comment (lucasdietrich) [reply tests/lib/json/src/main.c]: Indeed, reverted that

**Comment by carlescufi** [comment]
@mrfuchs please take a look

File: lib/os/json.c [hunk 4]
```diff
@@ -549,6 +540,8 @@ static int arr_parse(struct json_obj *obj,
 
 	if (val) {
 		elements = (size_t *)((char *)val + elem_descr->offset);
+	} else {
+		field = (char *)field + elem_descr->offset;
```
Comment (mrfuchs) [inline lib/os/json.c:544]: When the `field` pointer is updated and skipped over a (unused?) field, shouldn't the `last_elem` variable also be updated?

File: include/zephyr/sys/util.h [hunk 5]
```diff
@@ -210,6 +210,26 @@ extern "C" {
...
+ *
+ *	size_t size_of_baz = SIZEOF_FIELD(struct foo, baz);
+ *
+ * Above, @p size_of_baz is equal to sizeof(char).
+ *
+ * @param type the type of the structure containing the field
+ * @param member the name of the field in the structure
+ * @return the size of the field in bytes
+ */
+#define SIZEOF_FIELD(type, member) sizeof(((type *)0)->member)
```
Comment (mrfuchs) [inline include/zephyr/sys/util.h:231]: I have no strong opinion on this naming choice, but looking at the other macro names, shouldn't it be called `FIELD_SIZE` (cf. `ARRAY_SIZE`)?

**Review by mrfuchs — CHANGES_REQUESTED** [review]

File: include/zephyr/sys/util.h (hunk 5 above)
Comment (mrfuchs) [reply include/zephyr/sys/util.h:231]: Ah, sorry, I just saw that this names comes from @M1cha's advice with reference to the Linux kernel.

File: include/zephyr/data/json.h (hunk 2 above)
Comment (lucasdietrich) [reply include/zephyr/data/json.h]: Reverted the type for `sub_descr_len` to `size_t`.

I do agree, I moved the field `sub_struct_size` up and renamed it to `struct_size`. It just required a little more work to have all descriptors defined correctly: I needed to explicitly define the `struct_size` member to an arbitrary value (e.g. 0).

File: lib/os/json.c (hunk 4 above)
Comment (lucasdietrich) [reply lib/os/json.c:544]: In this new design `get_elem_size()` returns the actual size of the underlying object, so `last_elem` correctly reflects the address of the last element.

However I'm still not convinced the way `field` is adjusted when parsing the nested array (when parsing an array of array).

**Comment by lucasdietrich** [comment]
Did several modifications, but I still need to do additional testing.

**Comment by pdgendt** [comment]
The `align_shift` member becomes obsolete if the struct size is to be used.

File: include/zephyr/data/json.h [hunk 6]
```diff
@@ -106,6 +106,9 @@ struct json_obj_descr {
 	/* 65535 bytes is more than enough for many JSON payloads. */
 	uint32_t offset : 16;
 
+	/* Size of the structure being described by the "object" descriptor */
+	size_t struct_size;
```
Comment (pdgendt) [inline include/zephyr/data/json.h:110]: Remove the `align_shift` member. This frees up 2 bits, these can be used to `field_name_len` and `type` to get an extra bit?

File: include/zephyr/sys/util.h (hunk 5 above)
Comment (pdgendt) [inline include/zephyr/sys/util.h:231]: This should be added in a separate commit.

File: tests/lib/json/src/main.c [hunk 7]
```diff
@@ -11,6 +11,7 @@
 
 struct test_nested {
 	int nested_int;
+	uint32_t _unused_member;
```
Comment (pdgendt) [inline tests/lib/json/src/main.c:14]: Create a separate commit for the changes in tests.

**Comment by pdgendt** [comment]
@lucasdietrich care to update the PR for the latest comments, thanks

The tests in my PR #59485 succeed using this one.

**Comment by lucasdietrich** [comment]
Thank you for your review, I will make the changes this evening.

File: lib/os/json.c (hunk 4 above)
Comment (mrfuchs) [reply lib/os/json.c:544]: > However I'm still not convinced the way `field` is adjusted when parsing the nested array (when parsing an array of array).

Indeed. I hope that'll be fixed soon by #50816.

**Comment by lucasdietrich** [comment]
I believe I have addressed all your remarks @mrfuchs @pdgendt.

**Comment by pdgendt** [comment]
@lucasdietrich #50816 has been merged, please rebase and resolve conflicts, thanks!

**Comment by lucasdietrich** [comment]
Ok, i will rebase this evening.

**Comment by lucasdietrich** [comment]
Unfortunately I didn't have time to work on it yet, moreover the merge seems to be tricky.
I will need few more days.

**Comment by pdgendt** [comment]
@lucasdietrich any update on this?

**Comment by lucasdietrich** [comment]
After more investigation, I managed to get almost everything working except for the test "test_json_decoding_array_array", which has become a blocker for me.

I've somewhat lost track of the issue, but I can redirect my attention to it by tomorrow evening.

However, I might require some help to resolve it completely.

**Comment by lucasdietrich** [comment]
I managed to get almost everything working; we don't "guess" anymore the size of a structure but we compute it at build time and we use this to determine offsets.

The most tricky part was (and still is) the handling of nested arrays (array of arrays), which requires dedicated handling I'm not able to fully explain. To be honest I tried to have the unit tests working rather than designing an accurate solution for nested arrays.  This is where I would greatly benefit from external insight. Consequently this has left the following key issue unresolved : 

Yet, the member of the structure describing the nested array should remain at the start of the structure. Using the "test_json_encoding_array_array" test as example: I'm not yet able to calculate the correct offset of "objects" within the "struct array". Consequently the test fails (crashes) if we uncomment the line. Actually I don't fully understand when the adjustment should be made.

I've successfully tested these changes with a personnel project.

**Comment by lucasdietrich** [comment]
I've just rebased and noticed that the new tests introduced in #60046 are failing with the current changes(2 dim array encoding). I'll need more time to address this issue.
